modin/core/dataframe/algebra/binary.py

```diff
@@ -97,35 +136,100 @@ def compute_dtypes_common_cast(first, second) -> np.dtype:
+        If False is specified this method will return None if any of the operands doesn't
+        have materialized columns.
 
     Returns
     -------
-    dtypes
-        The pandas series with precomputed dtypes.
+    pandas.Series or None
+        The pandas series with precomputed dtypes or None if there's not enough metadata to compute it.
 
     Notes
     -----
     Finds a union of columns and finds dtypes for all these columns.
     """
+    if not trigger_computations:
+        if not first._modin_frame.has_columns_cache:
+            return None
+
+        if (
+            isinstance(second, type(first))
+            and not second._modin_frame.has_columns_cache
+        ):
+            return None
+
     columns_first = set(first.columns)
-    columns_second = set(second.columns)
-    columns_union = columns_first.union(columns_second)
-    dtypes = pandas.Series([np.dtype(bool)] * len(columns_union), index=columns_union)
-    dtypes = dtypes.sort_index()
+    if isinstance(second, type(first)):
+        columns_second = set(second.columns)
+        columns_union = columns_first.union(columns_second)
+    else:
+        columns_union = columns_first
+
+    dtypes = pandas.Series([dtype] * len(columns_union), index=columns_union)
+    return dtypes
+
+
+def try_compute_new_dtypes(first, second, infer_dtypes=None, result_dtype=None):
+    """
+    Precompute resulting dtypes of the binary operation if possible.
+
+    The dtypes won't be precomputed if any of the operands doesn't have their dtypes materialized
+    or if the second operand type is not supported. Supported types: PandasQueryCompiler, list,
+    dict, tuple, np.ndarray.
+
+    Parameters
+    ----------
+    first : PandasQueryCompiler
+        First operand of the binary operation.
+    second : PandasQueryCompiler, list-like or scalar
+        Second operand of the binary operation.
+    infer_dtypes : {"common_cast", "float", "bool", None}, default: None
+        How dtypes should be infered (see ``Binary.register`` doc for more info).
+    result_dtype : np.dtype, optional
+        NumPy dtype of the result. If not specified it will be inferred from the `infer_dtypes` parameter.
+
+    Returns
+    -------
+    pandas.Series or None
+    """
+    if infer_dtypes is None and result_dtype is None:
+        return None
+
+    try:
+        if infer_dtypes == "bool" or is_bool_dtype(result_dtype):
+            dtypes = maybe_build_dtypes_series(first, second, dtype=np.dtype(bool))
+        elif infer_dtypes == "common_cast":
+            dtypes = maybe_compute_dtypes_common_cast(first, second)
+        elif infer_dtypes == "float" or is_float_dtype(result_dtype):
+            dtypes = maybe_compute_dtypes_common_cast(first, second)
+            if dtypes is not None:
+                dtypes = dtypes.apply(coerce_int_to_float64)
+        elif result_dtype is not None:
+            dtypes = maybe_build_dtypes_series(first, second, dtype=result_dtype)
```

Comment:
It looks like this case is not being tested. Could you add a test?

Reply:
it seems that there's no way we can hit this branch for now, so I replaced it with a blank with an appropriate in-code comment